### DISCUSSION TITLE
how replace deprecated method Factory::getDbo()

### DISCUSSION THREAD

**Discussion by jschmi102** [joomla/joomla-cms#38111] [Q&A]
A well documented and widely used method (my systems shows more than 400 hits of it) to create a joomla database object is  **Factory::getDbo()** . 

With joomla 4 and (3.10?) this method is indicated as deprecated. I tried to find a well documented replacement but did not find any hints for it in the current joomla docs nor could i see any reasonable cause - besides not(?) to use statically functions to create db objects. Can anybody comment on this problem?

**Comment by MarkRS-UK** [comment]
The solution seems (to me) to be

`$db = Factory::getContainer()->get('DatabaseDriver');`

I found this by tracing to the getDbo() method of the Factory and seeing, after the deprecation method, what it does.
I've tested it, it works. I don't know if that's the official answer, though.

**Comment by jschmi102** [comment]
hi,
Though this is working, I'm still missing an official documentation for this change. And it is quite disapointing that an explanation or reason is missing why a commonly used and well documented method has been changed to be deprecated.

**Comment by xillibit** [comment]
Hello,

It seems there are removing static call, since Joomla! 4.2 you can use in some place : 
`$db = $this->getDatabase();`

**Comment by jschmi102** [comment]
hi,
is there any doc about this:  $db = $this->getDatabase();?

**Comment by jschmi102** [comment]
hi,
sofar I could not find any documentation or suggestions, how to use joomla's V4 new database interfaces. 

Joomla V4 is now out for quite a while and still there is are no infos about usage (e.g. of methods like "getDatabase()" or "getContainer()->get('DatabaseDriver');"). It is a bit frustrating that even samples of tutorials for joomla V4 still show and suggest the deprecated "getDbo()" method.

 Would be nice to read at least here what's going on...

**Comment by robbiejackson** [comment]
The deprecation notice says "Load the database service from the dependency injection container" and looking at Joomla 4.2.3 code the standard way seems to be:

    use Joomla\CMS\Factory;
    use Joomla\Database\DatabaseInterface;
    ...
    $db = Factory::getContainer()->get(DatabaseInterface::class);

However, a number of the Joomla classes use the `DatabaseAwareTrait` (defined in libraries/vendor/joomla/database/src/DatabaseAwareTrait.php in Joomla 4.2.3) to store the database object as a class instance variable - the trait basically does this and provides setter and getter methods. Within these classes (or subclasses thereof) it can therefore be accessed more efficiently by:

    $db = $this->getDatabase();

If you're developing a component using the standard Joomla MVC approach then you can use this approach within your model because in `MVCFactory::createModel()` (within libraries/src/MVC/Factory/MVCFactory.php) the code calls:

    $model->setDatabase($this->getDatabase());

However if you're developing module or plugin code then it doesn't set this up for you so have to use the first approach.

All that said, I totally agree with the sentiments regarding the lack of explanatory documentation :-(.
